plugins/vpc-tunnel/plugin/commands.go

```diff
@@ -0,0 +1,329 @@
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// Name templates used for objects created by this plugin.
+	GeneveLinkNameFormat = "gnv.%s%d"
+	bridgeNameFormat     = "gnvbr.%d"
+
+	// The default MAC address for the geneve interface can be any valid MAC address
+	// since there isn't an actual device as the gateway. The gateway IP address is
+	// just another IP from the same subnet as the primary IP address of the geneve interface.
+	defaultGatewayMacAddr = "00:aa:bb:cc:dd:ee"
+)
+
+// Add is the internal implementation of CNI ADD command.
+func (plugin *Plugin) Add(args *cniSkel.CmdArgs) error {
+	// Parse network configuration.
+	netConfig, err := tunnelConfig.New(args)
+	if err != nil {
+		log.Errorf("Failed to parse netconfig from args: %v.", err)
+		return err
+	}
+
+	log.Infof("Executing ADD with netconfig: %+v.", netConfig)
+
+	// Find the network namespace.
+	log.Infof("Searching for netns %s.", args.Netns)
+	ns, err := netns.GetNetNS(args.Netns)
+	if err != nil {
+		log.Errorf("Failed to find netns %s: %v.", args.Netns, err)
+		return err
+	}
+
+	// Create the Geneve object.
+	geneveName := fmt.Sprintf(GeneveLinkNameFormat, netConfig.VNI, netConfig.DestinationPort)
+	geneve, err := eni.NewGeneve(
+		geneveName,
+		netConfig.VNI,
+		netConfig.DestinationIPAddress,
+		netConfig.DestinationPort,
+		netConfig.Primary)
+	if err != nil {
+		log.Errorf("Failed to create geneve object %s: %v.", geneveName, err)
+		return err
+	}
+
+	// Create an interface for the geneve object.
+	log.Infof("Creating geneve link %s.", geneveName)
+	err = geneve.AttachToLink()
+	if err != nil {
+		if err.Error() == "device or resource busy" {
+			// If the geneve link already exists, it may have been created in a previous invocation
+			// of this plugin. Look for it in the target network namespace and reset it.
+			err = ns.Run(func() error {
+				err := geneve.ENI.AttachToLink()
+				if err != nil {
+					return err
+				}
+
+				// IP address and routing rules are only put in place for the interface
+				// only if it was the primary interface. So for non primary interface,
+				// we do not have to delete the same while reseting.
+				if !geneve.Primary {
+					return nil
+				}
+
+				hwAddr, err := net.ParseMAC(defaultGatewayMacAddr)
+				if err != nil {
+					log.Errorf("Failed to reset geneve link: %v", err)
+				}
+
+				neigh := &netlink.Neigh{
+					IP:           netConfig.GatewayIPAddress,
+					HardwareAddr: hwAddr,
+					LinkIndex:    geneve.GetLinkIndex(),
+				}
+				log.Infof("Deleting neighbour %+v.", neigh)
+				if err = netlink.NeighDel(neigh); err != nil {
+					log.Errorf("Failed to reset geneve link: %v", err)
+				}
+
+				route := &netlink.Route{
+					Gw:        netConfig.GatewayIPAddress,
+					LinkIndex: geneve.GetLinkIndex(),
+				}
+				log.Infof("Deleting default IP route %+v.", route)
+				err = netlink.RouteDel(route)
+				if err != nil {
+					if os.IsNotExist(err) || err.Error() == "no such process" {
```

Comment:
Can you add a comment for why we ignore "no such process" error?

Reply:
When a route is missing, an attempt to delete it will result in `no such process`. Removed `os.IsNotExist(err)` in following revision.